### PR TITLE
Prevent double click new issue/pull/comment button

### PR DESCRIPTION
**Pull request by a1012112796**
when network is not good, these button maybe
double clicked, then more than one same issues
pulls or comments will be created. this pull
request will fix this bug.

Before:
Peek 2021-06-15 11-57

Now:
Peek 2021-06-15 11-56

### CODE REVIEW

**Review by tomaswarynyca — APPROVED** [review]

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]:
```suggestion
  $btn.addClass('loading');
```

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]:
```suggestion
$(document).on('click', '.button.once-button.loading', (e) => {
```

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]:
```suggestion
```

**Comment by KN4CK3R** [comment]
Almost every button should become a `once-button`.

**Comment by silverwind** [comment]
I would call it `loading-button`.

**Comment by a1012112796** [comment]
> Almost every button should become a `once-button`.

Sure, all button which used to post a form by default logic can use this way. Will check more latter.

**Review by techknowlogick — APPROVED** [review]

**Comment by lunny** [comment]
Since almost button clicked will refresh the page, I think it's suitable. But once we convert to an ajax request button. It should be changed. @jolheiser has a similar PR #10459, but I think this one is better.

**Comment by a1012112796** [comment]
> Since almost button clicked will refresh the page, I think it's suitable. But once we convert to an ajax request button. It should be changed. @jolheiser has a similar PR #10459, but I think this one is better.

Hmm, ``timeout`` logic is usefull, I will check more later.

**Comment by zeripath** [comment]
CI failing

File: web_src/js/index.js
Comment (lunny) [inline web_src/js/index.js]: return

**Comment by lunny** [comment]
You can check the status code if it's 302 so that you could not change server side code.

File: web_src/js/index.js
Comment (a1012112796) [reply web_src/js/index.js]: ``reload`` is necessary.

**Comment by a1012112796** [comment]
> You can check the status code if it's 302 so that you could not change server side code.

It's hard because the browser will do something ...
ref:
> You can't handle redirects with XHR callbacks because the browser takes care of them automatically. You will only get back what at the redirected location.

http://stackoverflow.com/a/15996968

File: routers/web/repo/issue.go
Comment (jpraet) [inline routers/web/repo/issue.go]: How do things like

```
	if util.IsEmptyString(form.Title) {
		ctx.RenderWithErr(ctx.Tr("repo.issues.new.title_empty"), tplIssueNew, form)
		return
	}
```

on line 963 behave now that this form is submitted with AJAX?

File: routers/web/repo/issue.go
Comment (jpraet) [inline routers/web/repo/issue.go]:
```
						if issue.IsPull {
							ctx.Flash.Error(ctx.Tr("repo.issues.dependency.pr_close_blocked"))
							ctx.Redirect(fmt.Sprintf("%s/pulls/%d", ctx.Repo.RepoLink, issue.Index), http.StatusSeeOther)
						} else {
							ctx.Flash.Error(ctx.Tr("repo.issues.dependency.issue_close_blocked"))
							ctx.Redirect(fmt.Sprintf("%s/issues/%d", ctx.Repo.RepoLink, issue.Index), http.StatusSeeOther)
						}
```

does this still work now that this form is submitted with AJAX?

**Comment by zeripath** [comment]
Unfortunately this doesn't work correctly as I think some of the redirects are not correct.

Let's look at say create a pull request e.g.:

`http://localhost/gitea/administrator/merge-test/compare/master...administrator-patch-4`

When I go through the motions of creating the PR I get redirected back to the same URL which then says:

`A pull request between these branches already exists: administrator/merge-test#107`

If I double/triple click on the merge button I still end up with a failed merge too.

**Comment by lafriks** [comment]
Why not just disable button and add loading on it on click and be done with it?

I don't see need to change it to ajax requests etc

**Comment by a1012112796** [comment]
> Why not just disable button and add loading on it on click and be done with it?
> 
> I don't see need to change it to ajax requests etc

to add timeout and error handle logic ...

**Comment by a1012112796** [comment]
If no one like ajax way, I will revert later.

**Comment by lunny** [comment]
> Why not just disable button and add loading on it on click and be done with it?
> 
> I don't see need to change it to ajax requests etc

Ajax has a better UX than refresh the page.

**Comment by lafriks** [comment]
> Ajax has a better UX than refresh the page.

But it would still reload page after ajax request so there is not really much UX improvement in that

File: web_src/js/index.js
Comment (lafriks) [inline web_src/js/index.js]: It would be better to move this check to previous button click event and just check if button has class loading than prevent default and return

File: web_src/js/features/common-global.js
Comment (lafriks) [inline web_src/js/features/common-global.js]: Does return false prevent it to bubble event up? I think here it still needs e.preventDefault also

**Review by lafriks — APPROVED** [review]
Does this also will prevent with interacting with button twice using keyboard (sapcebar, enter)?

File: web_src/js/features/common-global.js
Comment (wxiaoguang) [inline web_src/js/features/common-global.js:153]: I think you need capture instead of bubble here. Otherwise the event handler on the button would be executed first (for example: ajax, IIRC some form buttons are doing ajax requests)

https://stackoverflow.com/questions/4616694/what-is-event-bubbling-and-capturing

File: web_src/js/features/common-global.js
Comment (wxiaoguang) [reply web_src/js/features/common-global.js]: Actually, `return false` in jQuery means preventDefault and stoppropagation at the same time.

Comment (lafriks) [reply web_src/js/features/common-global.js]: I don't know, that's why I asked 😅 have not really developed in jQuery for ages 🤣

File: web_src/js/features/common-global.js
Comment (wxiaoguang) [reply web_src/js/features/common-global.js:153]: To be clear, what I mean is:

If we are trying to introduce a general method, should the all cases be considered together? Will this mechanism be used for ajax buttons in future?

* If this mechanism is for traditional forms (non-ajax) only, these CSS names are too broad, it will be mis-used by ajax buttons in future.

* If this mechanism is for both traditional forms and ajax requests, it could be fine tuned some more.

----

update: I would suggest to use `form-once-submit` as the CSS name for these form-only buttons.

In future, we can have `ajax-once-request` CSS class for ajax buttons (if necessary), or ajax buttons should handle the repeated clicks by themselves.

Comment (lafriks) [reply web_src/js/features/common-global.js:153]: I think it should be generic for all use cases where technically possible and needed

**Comment by wxiaoguang** [comment]
> Does this also will prevent with interacting with button twice using keyboard (sapcebar, enter)?

It doesn't. These behaviors you said should be handled form's `submit` event.

File: web_src/js/features/common-global.js
Comment (wxiaoguang) [reply web_src/js/features/common-global.js:153]: Either is fine to me:

 * Make the CSS name in this PR clear for only traditional `form` buttons, and get this PR merged.
 * (OR) Let's spend some more time to fine tune some more.

How do you think?

**Review by lafriks — CHANGES_REQUESTED** [review]
it should be fixed to prevent button to submit form using keyboard

**Comment by wxiaoguang** [comment]
> it should be fixed to prevent button to submit form using keyboard

Oh sorry, maybe I didn't say clearly. To be clear:

* `enter` on the target button triggers `click`, it can be handled, see the demo : https://jsfiddle.net/L0r4yxtw/1/ , enter on the button won't submit the form.

* What I mean is, if user do enter on other text input fields in the form (not on the button), or use our `Ctrl+Enter` to submit, then the submit can not be handled. It is another case (should it be handled in this PR?)

Is it still a change request?

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Ok, handling form submit could be handled in other PR then imho

File: web_src/js/features/common-global.js
Comment (wxiaoguang) [inline web_src/js/features/common-global.js:155]:
```suggestion
    $btn.addClass('loading disabled');
    $btn.prop('disabled', true);
```

Should we use `$btn.prop('disabled', true);` to disable the button entirely? Then it's not clickable anymore, and we do not need to check `.loading` anymore, and no need to return false (preventDefault)

File: web_src/js/features/common-global.js
Comment (a1012112796) [reply web_src/js/features/common-global.js:153]: > Either is fine to me:
> 
> * Make the CSS name in this PR clear for only traditional `form` buttons, and get this PR merged.
> * (OR) Let's spend some more time to fine tune some more.
> 
> How do you think?

how about limit it by this way?
```suggestion
  // loading-button this logic used to prevent push one form more than one time
  // like double create new issue , pull request or comments
  $(document).on('click', 'form .button.loading-button', function (e) {
    const $btn = $(this);

    if ($btn.hasClass('loading')) {
      e.preventDefault();
      return false;
    }
```

Comment (wxiaoguang) [reply web_src/js/features/common-global.js:153]: `loading-button` is not a proper name ......

If you are reading some code, when you see a button called `loading-button`, you might think it is a button which is loading something.  BUT, the purpose of the class is to make the button can only submit the form once.

I would suggest to use a clear name to reduce misunderstanding.

File: web_src/js/features/common-global.js
Comment (a1012112796) [reply web_src/js/features/common-global.js:155]: looks semantic ui has handled it, just add `disabled` class is enough.

Comment (wxiaoguang) [reply web_src/js/features/common-global.js:155]: But what if it is not a Fomantic UI button? For example, what if some one writes a form without `.ui` class but uses your code?

If you require the developers to use your code in Fomantic UI context, you should change the selector to `.ui.button .xxxxxx` to make sure the code only works for Fomantic UI.

**Review by wxiaoguang — APPROVED** [review]
No blocker from my side.

**Comment by 6543** [comment]
:rocket:

**Pull request merged**
